src/polkit/polkitunixprocess.c

```diff
@@ -1131,6 +1361,437 @@ get_start_time_for_pid (pid_t    pid,
+#endif /* __linux__ */
+}
+
+static gchar *
+get_ctty_path_for_pidfd (gint     pidfd,
+                         GError **error)
+{
+#ifdef __linux__
+  guint tty_nr;
+  gint pid;
+  dev_t major;
+  dev_t minor;
+  struct stat st;
+  gchar *tty_name;
+  gchar *filename;
+  gchar *contents;
+  size_t length;
+  gchar **tokens;
+  guint num_tokens;
+  gchar *p;
+  gchar *endp;
+
+  pid = 0;
+  tty_name = NULL;
+  contents = NULL;
+  tokens = NULL;
+
+  g_return_val_if_fail (pidfd >= 0, NULL);
+
+  pid = polkit_unix_process_get_pid_from_pidfd (pidfd, error);
+  if (pid <= 0)
+    return NULL;
+
+  filename = g_strdup_printf ("/proc/%d/stat", pid);
+  if (!filename)
+    {
+      g_set_error (error,
+                   POLKIT_ERROR,
+                   POLKIT_ERROR_FAILED,
+                   "Error allocating memory for procfs path");
+      goto out;
+    }
+
+  if (!g_file_get_contents (filename, &contents, &length, error))
+    goto out;
+
+  /* parent's pid is the token at index 1 after the '(process name)' entry - since only this
+   * field can contain the ')' character, search backwards for this to avoid malicious
+   * processes trying to fool us
+   */
+  p = strrchr (contents, ')');
+  if (p == NULL)
+    {
+      g_set_error (error,
+                   POLKIT_ERROR,
+                   POLKIT_ERROR_FAILED,
+                   "Error parsing file %s",
+                   filename);
+      goto out;
+    }
+  p += 2; /* skip ') ' */
+  if (p - contents >= (int) length)
+    {
+      g_set_error (error,
+                   POLKIT_ERROR,
+                   POLKIT_ERROR_FAILED,
+                   "Error parsing file %s",
+                   filename);
+      goto out;
+    }
+
+  tokens = g_strsplit (p, " ", 0);
+
+  num_tokens = g_strv_length (tokens);
+
+  if (num_tokens < 5)
+    {
+      g_set_error (error,
+                   POLKIT_ERROR,
+                   POLKIT_ERROR_FAILED,
+                   "Error parsing file %s",
+                   filename);
+      goto out;
+    }
+
+  /*
+   * The string can have negative representation, but it's actually an unsigned int,
+   * cast it to convert it
+   */
+  tty_nr = (guint) strtoll (tokens[4], &endp, 10);
+  if (endp == tokens[4])
+    {
+      g_set_error (error,
+                   POLKIT_ERROR,
+                   POLKIT_ERROR_FAILED,
+                   "Error parsing file %s",
+                   filename);
+      goto out;
+    }
+
+  if (tty_nr == 0)
+    {
+      g_set_error (error,
+                   POLKIT_ERROR,
+                   POLKIT_ERROR_FAILED,
+                   "No TTY associated with process %d",
+                   pid);
+      goto out;
+    }
+
+  /*
+   * As per the manpage, the minor device number is contained in the
+   * combination of bits 31 to 20 and 7 to 0, the major device number is in
+   * bits 15 to 8.
+   */
+  major = (tty_nr & 0xff00) >> 8;
+  minor = ((tty_nr & 0xfff00000) >> 12) | (tty_nr & 0xff);
+
+  if (major < 136 || major > 143)
```

Comment:
uh, this is ugly. why do this at all? if something shows up as ctty for a process then it's a tty, no need to validate that. just trust it?

if you want to properly check if something is a pty on linux, better look if backing fs is devpts, and if so check how it reacts to TIOCGPKT. See how terminal_is_pty_fd() does it in the systemd codebase.

but again: why restrict this to pty here? any tty should be fine.

and also: don't bother with ttys here at all, use getsid()/setsid() if anthing.

Reply:
> uh, this is ugly. why do this at all? if something shows up as ctty for a process then it's a tty, no need to validate that. just trust it?

This crosses a security boundary, ie we are inspecting metadata of a foreign, untrusted process. So I thought it was a good idea to dot the Is and cross the Ts

Reply:
don't second guess the kernel. 

i don't get you philosophy on this. one one hand you don't trust the kernel enough to make only ttys cttys, and second guess it hence. on the other hand you yolo around races...

Reply:
Already removed, you might need to refresh the tab if GH still shows this